src/streams/tls.c

```diff
@@ -5,41 +5,124 @@
  * a Linking Exception. For full terms see the included COPYING file.
  */
 
-#include "streams/tls.h"
-
 #include "git2/errors.h"
 
+#include "common.h"
+#include "global.h"
+#include "streams/tls.h"
 #include "streams/mbedtls.h"
 #include "streams/openssl.h"
 #include "streams/stransport.h"
 
-static git_stream_cb tls_ctor;
+struct git_tls_stream_registration {
+	git_rwlock lock;
+	git_stream_registration callbacks;
+};
+
+static struct git_tls_stream_registration stream_registration;
+
+static void shutdown_ssl(void)
+{
+	git_rwlock_free(&stream_registration.lock);
+}
+
+int git_tls_stream_global_init(void)
+{
+	if (git_rwlock_init(&stream_registration.lock) < 0)
+		return -1;
+
+	git__on_shutdown(shutdown_ssl);
+	return 0;
+}
 
-int git_stream_register_tls(git_stream_cb ctor)
+int git_stream_register_tls(git_stream_registration *registration)
 {
-	tls_ctor = ctor;
+	assert(!registration || registration->init);
 
+	if (git_rwlock_wrlock(&stream_registration.lock) < 0) {
```

Comment:
This is something we support today and its relied on in LibGit2Sharp.  Since LibGit2Sharp itself supports pluggable streams, we have tests that create a stream and invoke it.  Those tests might get invoked at any time, not necessarily first-thing (nor is their lifecycle until the end of the test runs).